### PR TITLE
Refactor CI

### PR DESCRIPTION
**Pull request by JoaoRodrigues**
Changes to workflow:
- Replace CodeCov by simple coverage report/check
- Build/Coverage for every OS/Python version.
- Try and reinstate windows testing.

### CODE REVIEW

**Comment by JoaoRodrigues** [comment]
Some more changes:
- Added `unittest.skipIf` to tests requiring a TTY, so that they are skipped on windows which is where we cannot have a workaround.

**Comment by JoaoRodrigues** [comment]
@joaomcteixeira, have a look and give your opinion. I tried making this a consensus of our opinions :)

**Comment by JoaoRodrigues** [comment]
By the way, the workflows were not running because of a syntax issue with YAML. So, I added a separate YAML linter to detect these cases. It's a bit of a chicken and egg problem (since the workflow to detect yaml issues is written in yaml) but since we start from a 'clean' slate of working yaml files, it should be easy to catch future failures.

**Comment by joaomcteixeira** [comment]
* I don't mind having the report sent to Codecov. I think it facilitates newcomers (and us) to visualize the coverage rate quickly. It also integrates with the badge in the readme. However, there is no real need for it, as we can see the coverage rates in the GitHub action, though it requires some clicks and knowledge. As you said, running CIs everywhere has its energy consumption penalty, but running another build also does. Honestly, I don't know which is more energy efficient: sending the coverage reports to codecov after each testing or merging them at the end in the Github actions. I am OK with either way here. If we drop Codecov, also remove the badge in the README.

* The Yaml linter is chicken and egg, indeed. Usually, actions won't trigger if there's a syntax error in the YAML. But still, it is a good addition.

* I would add a small comment in the `setup.cfg` or in the `CONTRIBUTING` with the commands developers need to run locally for flake8 and coverage. So developers can access lint results and the coverage table before PR. Before, we were not checking lint in the CIs, but if this PR is accepted, we will; so developers need to have local access to it.

* Regarding skipping tests on windows. I suggest using a global variable instead of ignoring the tests when `sys.platform.startswith('win')`. I suggest we define a global variable in the YAML CI and then `@unittest.skipIf("VARIABLE" in os.environ)`. In this way, WIN users would have access to all tests locally, and tests would only be skipped in the Actions.

**Comment by JoaoRodrigues** [comment]
>I don't mind having the report sent to Codecov. I think it facilitates newcomers (and us) to visualize the coverage rate quickly. It also integrates with the badge in the readme. However, there is no real need for it, as we can see the coverage rates in the GitHub action, though it requires some clicks and knowledge. As you said, running CIs everywhere has its energy consumption penalty, but running another build also does. Honestly, I don't know which is more energy efficient: sending the coverage reports to codecov after each testing or merging them at the end in the Github actions. I am OK with either way here. If we drop Codecov, also remove the badge in the README.

👍 

>I would add a small comment in the setup.cfg or in the CONTRIBUTING with the commands developers need to run locally for flake8 and coverage. So developers can access lint results and the coverage table before PR. Before, we were not checking lint in the CIs, but if this PR is accepted, we will; so developers need to have local access to it.

Noted, will update the `CONTRIBUTING` file.

>Regarding skipping tests on windows. I suggest using a global variable instead of ignoring the tests when sys.platform.startswith('win'). I suggest we define a global variable in the YAML CI and then @unittest.skipIf("VARIABLE" in os.environ). In this way, WIN users would have access to all tests locally, and tests would only be skipped in the Actions.

Good point. I can have a check for a specific env variable instead.

File: .github/workflows/build_and_test.yml [hunk 1]
```diff
@@ -1,164 +1,215 @@
...
+          bump2version --dry-run --verbose --allow-dirty major
+
+      - name: Test
+        run: |
+          python -m pip install .
+          script -q -e -c "coverage run -p setup.py test"
+
+      # Store coverage data for later
+      # https://hynek.me/articles/ditch-codecov-python/
+      - name: Upload coverage data
```
Comment (joaomcteixeira) [inline .github/workflows/build_and_test.yml:81]: This upload is no longer needed since we dropped codecov. Or is it? :thinking:

**Review by joaomcteixeira — CHANGES_REQUESTED** [review]
Great work! Some changes are still needed:

File: tests/test_pdb_selatom.py [hunk 2]
```diff
@@ -94,6 +94,7 @@ def test_file_not_found(self):
         self.assertEqual(self.stderr[0][:22],
                          "ERROR!! File not found")  # proper error message
 
+    @unittest.skipIf(os.getenv('GITHUB_ACTIONS'), 'skip on GHA windows - no TTY')
```
[the remark this answers is not in the document]
Comment (JoaoRodrigues) [reply tests/test_pdb_selatom.py:97]: Gah. We can actually skip all of them in all OSes and avoid the silly repetition of all the OSes, what do you think? These are only two tests: file missing and help text.

File: .github/workflows/build_and_test.yml (hunk 1 above)
Comment (joaomcteixeira) [reply .github/workflows/build_and_test.yml:81]: And does it need to do that combine for every OS and python version or only for all of them in the end?

Comment (JoaoRodrigues) [reply .github/workflows/build_and_test.yml:81]: It needs to upload the artifact at the end of every run, then the combination needs to grab the artifacts and combine them in one step. I should use 'store' actually, although upload is correct as well. I thought this was in another place in the code when I suggested 'combine'. Basically, it's uploading from the runner to a centralized location that can persist and be accessed by other future runners.

Comment (joaomcteixeira) [reply .github/workflows/build_and_test.yml:81]: Okay, now you said it differently. If ignoring those tests makes the YAML collapse to a single matrix, then my vote. You are right there are some commands OS-specific for the different builds to workaround the TTY. Can we make it to run the TTY at least on ubuntu, or not even?

File: tests/test_pdb_selatom.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (joaomcteixeira) [reply tests/test_pdb_selatom.py:97]: Okay, now you said it differently. If ignoring those tests makes the YAML collapse to a single matrix, then my vote. You are right there are some commands OS-specific for the different builds to workaround the TTY. Can we make it to run the TTY at least on ubuntu, or not even?

File: .github/workflows/build_and_test.yml (hunk 1 above)
Comment (joaomcteixeira) [reply .github/workflows/build_and_test.yml:81]: Thanks for explaining :+1:

File: tests/test_pdb_selatom.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (JoaoRodrigues) [reply tests/test_pdb_selatom.py:97]: That's easier.. it's an if statement.

[the remark this answers is not in the document]
Comment (JoaoRodrigues) [reply tests/test_pdb_selatom.py:97]: CI: Continuous Irritation :)

[the remark this answers is not in the document]
Comment (joaomcteixeira) [reply tests/test_pdb_selatom.py:97]: :rofl: you nailed it!

**Comment by joaomcteixeira** [comment]
Perfect end result! :+1: really nice. :clap: after so many commits ... feels good :feelsgood:

**Pull request merged**